test/model/message_test.dart

```diff
@@ -76,6 +94,341 @@ void main() {
+        await prepareOutboxMessageToFailAfterDelay(
+          kSendMessageOfferRestoreWaitPeriod + Duration(seconds: 1));
+        async.elapse(kLocalEchoDebounceDuration);
+        checkState().equals(OutboxMessageState.waiting);
+
+        async.elapse(kSendMessageOfferRestoreWaitPeriod - kLocalEchoDebounceDuration);
+        checkState().equals(OutboxMessageState.waitPeriodExpired);
+
+        // Handle the message event while the message is being sent.
+        await store.handleEvent(eg.messageEvent(message,
+          localMessageId: store.outboxMessages.keys.single));
+        check(store.outboxMessages).isEmpty();
+
+        // Complete the send request.  There should be no error despite
+        // the send request failure, because the outbox message is not
+        // in the store any more.
+        await check(outboxMessageFailFuture).completes();
+      }));
+
+      test('waiting -> waitPeriodExpired -> (delete) because outbox message was taken', () => awaitFakeAsync((async) async {
+        // Set up an error to fail `sendMessage` with a delay, leaving time for
+        // the outbox message to be taken (by the user, presumably).
+        await prepareOutboxMessageToFailAfterDelay(
+          kSendMessageOfferRestoreWaitPeriod + Duration(seconds: 1));
+        async.elapse(kLocalEchoDebounceDuration);
+        checkState().equals(OutboxMessageState.waiting);
+
+        async.elapse(kSendMessageOfferRestoreWaitPeriod - kLocalEchoDebounceDuration);
+        checkState().equals(OutboxMessageState.waitPeriodExpired);
+
+        store.takeOutboxMessage(store.outboxMessages.keys.single);
+        check(store.outboxMessages).isEmpty();
+      }));
+
+      test('failed -> (delete) because event received', () => awaitFakeAsync((async) async {
+        await prepareOutboxMessageToFailAfterDelay(Duration.zero);
+        await check(outboxMessageFailFuture).throws();
+        checkState().equals(OutboxMessageState.failed);
+
+        await store.handleEvent(eg.messageEvent(message,
+          localMessageId: store.outboxMessages.keys.single));
+        check(store.outboxMessages).isEmpty();
+      }));
+
+      test('failed -> (delete) because outbox message was taken', () => awaitFakeAsync((async) async {
+        await prepareOutboxMessageToFailAfterDelay(Duration.zero);
+        await check(outboxMessageFailFuture).throws();
+        checkState().equals(OutboxMessageState.failed);
+
+        store.takeOutboxMessage(store.outboxMessages.keys.single);
+        check(store.outboxMessages).isEmpty();
+      }));
+    });
+
+    test('when sending to "(no topic)", process topic like the server does when creating outbox message', () => awaitFakeAsync((async) async {
+      await prepareOutboxMessage(
+        destination: StreamDestination(stream.streamId, TopicName('(no topic)')),
+        zulipFeatureLevel: 370);
+      async.elapse(kLocalEchoDebounceDuration);
+      check(store.outboxMessages.values.single)
```

Comment:
nit: use checks-getters
```suggestion
      check(store.outboxMessages).values.single
```